crates/ruma-server-util/src/authorization.rs

```diff
@@ -0,0 +1,198 @@
+            debug!("Field destination duplicated in X-Matrix Authorization header");
+        }
+        self.destination = Some(destination);
+        self
+    }
+
+    /// Set the key attribute
+    pub fn key(mut self, key: Box<ServerSigningKeyId>) -> Self {
+        if self.key.is_some() {
+            debug!("Field key duplicated in X-Matrix Authorization header");
+        }
+        self.key = Some(key);
+        self
+    }
+
+    /// Set the sig attribute
+    pub fn sig(mut self, sig: String) -> Self {
+        if self.sig.is_some() {
+            debug!("Field sig duplicated in X-Matrix Authorization header");
+        }
+        self.sig = Some(sig);
+        self
+    }
+
+    /// Build an [`XMatrix`] instance
+    pub fn build(self) -> Option<XMatrix> {
+        Some(XMatrix {
+            origin: self.origin?,
+            destination: self.destination,
+            key: self.key?,
+            sig: self.sig?,
+        })
+    }
+}
+
+impl Default for XMatrixBuilder {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Credentials for XMatrix {
+    const SCHEME: &'static str = "X-Matrix";
+
+    fn decode(value: &HeaderValue) -> Option<Self> {
+        debug_assert!(
+            value.as_bytes().starts_with(b"X-Matrix "),
+            "HeaderValue to decode should start with \"X-Matrix ..\", received = {:?}",
+            value,
+        );
+
+        let parameters =
+            std::str::from_utf8(&value.as_bytes()["X-Matrix ".len()..]).ok()?.trim_start();
+
+        let mut builder: XMatrixBuilder = Default::default();
+
+        for entry in parameters.split_terminator(',') {
+            let (name, value) = entry.split_once('=')?;
+
+            // It's not at all clear why some fields are quoted and others not in the spec,
+            // let's simply accept either form for every field.
+            let value =
+                value.strip_prefix('"').and_then(|rest| rest.strip_suffix('"')).unwrap_or(value);
+
+            builder = match name {
+                "origin" => builder.origin(value.try_into().ok()?),
+                "destination" => builder.destination(value.try_into().ok()?),
+                "key" => builder.key(value.try_into().ok()?),
+                "sig" => builder.sig(value.to_owned()),
+                name => {
+                    debug!("Unknown field {} found in X-Matrix Authorization header", name);
+                    builder
+                }
+            };
+        }
+
+        builder.build()
+    }
+
+    fn encode(&self) -> HeaderValue {
+        if let Some(destination) = &self.destination {
+            format!(
+                "X-Matrix origin=\"{}\",destination=\"{}\",key=\"{}\",sig=\"{}\"",
+                self.origin, destination, self.key, self.sig
+            )
+        } else {
+            format!("X-Matrix origin=\"{}\",key=\"{}\",sig=\"{}\"", self.origin, self.key, self.sig)
+        }
+        .try_into()
+        .expect("")
```

Comment:
Please add a descriptive panic message, even if it is "header format is statically determined"